### PR TITLE
api: postPolicy cleanup. Simplify the code and re-use.

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
api: postPolicy cleanup. Simplify the code and re-use.

## Motivation and Context
This change is cleanup of the postPolicyHandler code
primarily to address the flow and also converting
certain critical parts into self contained functions.

## How Has This Been Tested?
Manually, mc and `go test`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/handler-utils.go [hunk 1]
```diff
@@ -126,51 +130,67 @@ func extractMetadataFromHeader(header http.Header) map[string]string {
...
-		// Form field names are case insensitive, look for both canonical
-		// and non canonical entries.
-		if _, ok := formValues[canonicalHeader]; ok {
-			metadata[supportedHeader] = formValues[canonicalHeader]
-		} else if _, ok := formValues[supportedHeader]; ok {
-			metadata[supportedHeader] = formValues[canonicalHeader]
-		}
+// The Query string for the redirect URL the client is
+// redirected on successful upload.
+func redirectPostRawQuery(objInfo ObjectInfo) string {
```
Comment (balamurugana) [inline cmd/handler-utils.go]: You could name this function as `GetPostRawQuery()` or `GetRedirectPostRawQuery()`

File: cmd/handler-utils.go [hunk 2]
```diff
@@ -126,51 +130,67 @@ func extractMetadataFromHeader(header http.Header) map[string]string {
...
+	redirectValues := make(url.Values)
+	redirectValues.Set("bucket", objInfo.Bucket)
+	redirectValues.Set("key", objInfo.Name)
+	redirectValues.Set("etag", "\""+objInfo.MD5Sum+"\"")
```
Comment (balamurugana) [inline cmd/handler-utils.go]: You could use `` `"` ``

File: cmd/handler-utils.go (hunk 1 above)
Comment (vadmeste) [reply cmd/handler-utils.go]: or `genRedirectPostRawQuery()` :)

Comment (vadmeste) [reply cmd/handler-utils.go]: LGTM otherwise

Comment (harshavardhana) [reply cmd/handler-utils.go]: Done

File: cmd/handler-utils.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/handler-utils.go]: Done

File: cmd/bucket-handlers.go [hunk 3]
```diff
@@ -439,14 +436,25 @@ func (api objectAPIHandlers) PostPolicyBucketHandler(w http.ResponseWriter, r *h
...
 	bucket := mux.Vars(r)["bucket"]
-	formValues["Bucket"] = bucket
+	formValues.Set("Bucket", bucket)
 
-	if fileName != "" && strings.Contains(formValues["Key"], "${filename}") {
+	if fileName != "" && strings.Contains(formValues.Get("Key"), "${filename}") {
 		// S3 feature to replace ${filename} found in Key form field
 		// by the filename attribute passed in multipart
-		formValues["Key"] = strings.Replace(formValues["Key"], "${filename}", fileName, -1)
+		formValues["Key"] = []string{strings.Replace(formValues.Get("Key"), "${filename}", fileName, -1)}
```
Comment (vadmeste) [inline cmd/bucket-handlers.go]: formValues.Set("Key", ..) instead?

Comment (harshavardhana) [reply cmd/bucket-handlers.go]: Makes sense..

**Review by vadmeste — APPROVED** [review]

**Pull request merged**